[SDK-3295] Add Spring Interactive Quickstart (#10087)
auth0/docs: articles/quickstart/backend/java-spring-security5/interactive.md

---
title: Spring Boot API
description: Secure your API using Spring Security 5 and Auth0
interactive: true
alias:
  - spring security
  - spring
topics:
  - quickstart
  - backend
  - spring
files:
  - files/application
  - files/audience-validator
  - files/security-config
  - files/message
  - files/api-controller
github:
  path: 01-Authorization-MVC
---

# Secure your API using Spring Security 5 and Auth0

> Auth0 allows you to quickly add authentication and gain access to user profile information in your application. This guide demonstrates how to integrate Auth0 with any new or existing Spring Boot application using the [auth0-spring-security-api](https://github.com/auth0/auth0-spring-security-api) package.

Reviewer: Is it expected that this SDK isn't installed as a dependency or used in any of the code snippets?

Author: I am not sure what we are expecting here Fred, we already have a code snippet that shows how to add the dependency

> Auth0 allows you to quickly add authentication and gain access to user profile information in your application. This guide demonstrates how to integrate Auth0 with any new or existing Spring Boot application using the [auth0-spring-security-api](https://github.com/auth0/auth0-spring-security-api) package.

Reviewer: Should we talk about Authorization here instead of Authentication and User Profile Information?

Author: We are not doing that here Fred because the authorisation functionality is currently provided by the spring framework itself `.hasAuthority("SCOPE_read:messages")`. Auth0 provides scope for authorisation and Spring framework does the authorisation. Our SDK doesn't play a role here

Reviewer: What I mean is

> Auth0 allows you to quickly add authentication and gain access to user profile information in your application. 

Should probably be

> Auth0 allows you to quickly add authorization to your application. 

In an API scenario, we talk about Authorization, not Authentication.

If you haven't created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with.

Alternatively, you can read [our getting started guide](get-started/auth0-overview/set-up-apis) that helps you set up your first API through the Auth0 dashboard.

Every API in Auth0 is configured using an API Identifier that your application code will use as the Audience to validate the access token.

<%= include('../../../_includes/_api_auth_intro') %>

## Define permissions
<%= include('../_includes/_api_scopes_access_resources') %>

> ## Configure the Sample Project {{{ data-action=code data-code="application.yml#1:8" }}}

Reviewer: ```suggestion
## Configure the sample project {{{ data-action=code data-code="application.yml#1:8" }}}
```

    <artifactId>spring-boot-starter-parent</artifactId>
    <version>2.5.12</version>
    <relativePath/>
</parent>

<dependencies>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
    </dependency>
</dependencies>
```

## Validate the audience {{{ data-action=code data-code="AudienceValidator.java#13:20" }}}

To validate the JWT, you also need to validate that the JWT is intended for your API by checking the `aud` claim of the JWT. Create a new class named `AudienceValidator` that implements the `OAuth2TokenValidator` interface and override the `validate` method to verify whether the required `audience` is present.

## Configure the resource server {{{ data-action=code data-code="SecurityConfig.java" }}}

To configure the application as a Resource Server and validate the JWTs, create a class that extends `WebSecurityConfigurerAdapter`, add the `@EnableWebSecurity` annotation, and override the `configure` method

> ### Protect API Endpoints

Reviewer: ```suggestion
### Protect API endpoints
```

<%= include('../_includes/_api_endpoints') %>

The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:

::: note
By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.

If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) to learn how to customize the extracted authorities.
:::

### Configure JWT Validator

Update the `SecurityConfig` class to configure a `JwtDecoder` bean that uses the `AudienceValidator`. The `auth0.audience` value from `application.yml` will be used to validate the `aud` claim. The `issuer` is verified using the `issuer-uri` value from the `application.yml`

> ## Create the Domain Object {{{ data-action=code data-code="Message.java#1:11" }}}

Reviewer: I'm curious if everyone knows what is meant by a Domain Object. Also curious if this is something we should explicitly have in our quickstart? Would it make sense to just return an anonymous class or a String instead of an instance of Message from the controller? Avoids the need for terminology like Domain Object, while also avoids solving the issue mentioned in the short sentence below.

Author: We have fixed this and also created a sample which just returns a string instead of a POJO - https://github.com/auth0-samples/auth0-spring-security5-api-sample/pull/16

Author: Unfortunately we won't be able to remove this :( because all API based quickstarts use a test script and we need to return a JSON object instead of a string. We cannot edit the test script since all the other QS samples depend on it - https://github.com/search?q=org%3Aauth0-samples++api-quickstarts-tests&type=code

> Create a new class named `Message`. This will act as a simple domain object to return during the API calls.

Reviewer: This is a pretty short sentence, which might result in some issues while scrolling, as the Message.java file will only be shown for only a small time while scrolling. Is there any way we can improve this here?

Author: We have removed this by avoiding creation of the POJO. Though the use of POJO returns a proper JSON, for the sake of sample, we will just return a String

Author: We cannot remove this as explained [here](https://github.com/auth0/docs/pull/10087#discussion_r883553502) so we have added more description to this section to avoid scroll issues

## Create the API controller {{{ data-action=code data-code="APIController.java" }}}

Create a new class named `APIController` to handle requests to the endpoints. The `APIController` will have three routes as defined in the [Protect API Endpoints](/quickstart/backend/java-spring-security5/interactive/#configure-the-resource-server) section. For the simplicity of this sample we will allow all origins through `@CrossOrigin` annotation. Real applications should configure `CORS` for their use case.

> ## Run the Application {{{ data-action=code data-code="APIController.java" }}}

Reviewer: ```suggestion
## Run the application {{{ data-action=code data-code="APIController.java" }}}
```

To build and run the sample project, execute the `bootRun` Gradle task.

Linux or macOS:

```bash
./gradlew bootRun
```

Windows:

```bash
gradlew.bat bootRun
```

If you are configuring your own application using Maven and the [Spring Boot Maven Plugin](https://docs.spring.io/spring-boot/docs/current/reference/html/build-tool-plugins-maven-plugin.html), you can execute the `spring-boot:run` goal.

Linux or macOS:

```bash
mvn spring-boot:run
```

Windows:

```bash
mvn.cmd spring-boot:run
```

> The sample application will be available at `http://localhost:3010/`. Read about how to test and use your API in the [Using Your API](/quickstart/backend/java-spring-security5/02-using) article. Incase of unexpected responses from your API, use the [Troubleshooting](/quickstart/backend/java-spring-security5/03-troubleshooting) section to check your configuration.

Reviewer: Can we add a checkpoint here ? For example as done here https://github.com/auth0/docs/pull/10085

---
title: Spring Boot API
description: Secure your API using Spring Security 5 and Auth0
interactive: true
alias:
  - spring security
  - spring
topics:
  - quickstart
  - backend
  - spring
files:
  - files/application
  - files/audience-validator
  - files/security-config
  - files/message
  - files/api-controller
github:
  path: 01-Authorization-MVC
---

# Secure your API using Spring Security 5 and Auth0

> Auth0 allows you to quickly add authorization to your application. This guide demonstrates how to integrate Auth0 with any new or existing Spring Boot application.

Reviewer 2: ```suggestion
Auth0 allows you to add authentication and gain access to user profile information in your application. This guide demonstrates how to integrate Auth0 with any new or existing Spring Boot application.
```

Reviewer: Is it intentional that we are adding back the reference to the SDK here? I believe this got removed because this Quickstart doesn't use that specific SDK. Can you verify @poovamraj  ?

Also curious why we want to change Authorization to Authentication and User Profile information again? This Quickstart is about adding Authorization to an API, not so much about Authentication or User Profile info.

Reviewer 2: Updated.

> If you haven't created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with.

Reviewer 2: ```suggestion
If you have not created an API in your Auth0 dashboard yet, use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with.
```

> Alternatively, you can read [our getting started guide](get-started/auth0-overview/set-up-apis) that helps you set up your first API through the Auth0 dashboard.

Reviewer 2: ```suggestion
Review [our getting started guide](get-started/auth0-overview/set-up-apis) to set up your first API through the Auth0 dashboard.
```

> Every API in Auth0 is configured using an API Identifier that your application code will use as the Audience to validate the access token.

Reviewer 2: ```suggestion
Each Auth0 API uses the API Identifier, which your application needs to validate the access token.
```

<%= include('../../../_includes/_api_auth_intro') %>

## Define permissions
<%= include('../_includes/_api_scopes_access_resources') %>

## Configure the sample project {{{ data-action=code data-code="application.yml#1:8" }}}

> The sample project uses a `/src/main/resources/application.yml` file, which configures it to use the correct Auth0 **domain** and **API Identifier** for your API. If you download the code from this page it will be automatically configured. If you clone the example from GitHub, you will need to fill it in yourself.

Reviewer 2: ```suggestion
The sample project uses a `/src/main/resources/application.yml` file, which configures it to use the correct Auth0 **domain** and **API Identifier** for your API. The example available for download on this page is automatically configured. If you choose to clone the example from GitHub, you need to configure it yourself.
```

Reviewer: I am wondering if leaving something that references "download" would make sense? Because it actually is about the example that you can download, would it confuse people if we remove the "download" part in the sense that they might not know what example we are referring to?

Reviewer 2: Updated.

| Attribute | Description|
| --- | --- |
| `auth0.audience` | The unique identifier for your API. If you are following the steps in this tutorial it would be `https://quickstarts/api`. |

> | `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The issuer URI of the resource server, which will be the value of the `iss` claim in the JWT issued by Auth0. Spring Security will use this property to discover the authorization server's public keys and validate the JWT signature. The value will be your Auth0 domain with an `https://` prefix and a `/` suffix (the trailing slash is important). |

Reviewer 2: ```suggestion
| `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The value of the `iss` claim and the issuer URI of the resource server in the JWT issued by Auth0. Spring Security uses this property to discover the authorization server's public keys and validate the JWT signature. The value is Auth0 domain with an `https://` prefix and a `/` suffix (the trailing slash is important). |
```

    <version>2.5.12</version>
    <relativePath/>
</parent>

<dependencies>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
    </dependency>
</dependencies>
```

## Validate the audience {{{ data-action=code data-code="AudienceValidator.java#13:20" }}}

To validate the JWT, you also need to validate that the JWT is intended for your API by checking the `aud` claim of the JWT. Create a new class named `AudienceValidator` that implements the `OAuth2TokenValidator` interface and override the `validate` method to verify whether the required `audience` is present.

## Configure the resource server {{{ data-action=code data-code="SecurityConfig.java" }}}

To configure the application as a Resource Server and validate the JWTs, create a class that extends `WebSecurityConfigurerAdapter`, add the `@EnableWebSecurity` annotation, and override the `configure` method

### Protect API endpoints

<%= include('../_includes/_api_endpoints') %>

> The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:

Reviewer 2: ```suggestion
The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers restrict access based on the level of authorization required:
```

::: note

> By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.

Reviewer 2: ```suggestion
By default, Spring Security creates a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This scope enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.
```

If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) to learn how to customize the extracted authorities.
:::

### Configure JWT Validator

> Update the `SecurityConfig` class to configure a `JwtDecoder` bean that uses the `AudienceValidator`. The `auth0.audience` value from `application.yml` will be used to validate the `aud` claim. The `issuer` is verified using the `issuer-uri` value from the `application.yml`

Reviewer 2: ```suggestion
Update the `SecurityConfig` class to configure a `JwtDecoder` bean that uses the `AudienceValidator`. The `auth0.audience` value from `application.yml` validates the `aud` claim. The `issuer-uri` value from `application.yml` verifies the `issuer`.
```

## Create the Domain Object {{{ data-action=code data-code="Message.java#1:11" }}}

> To make your endpoint return a JSON, you can use a [POJO](https://en.wikipedia.org/wiki/Plain_old_Java_object) (Plain Old Java Object). The member variables of this object is serialized into the key value for your JSON. Create a new class named `Message`. This will act as a simple domain object to return during the API calls.

Reviewer 2: ```suggestion
To make your endpoint return a JSON, you can use a [POJO](https://en.wikipedia.org/wiki/Plain_old_Java_object) (Plain Old Java Object). The member variables of this object is serialized into the key value for your JSON. Create a new class named `Message` as an example domain object to return during the API calls.
```

## Create the API controller {{{ data-action=code data-code="APIController.java" }}}

> Create a new class named `APIController` to handle requests to the endpoints. The `APIController` will have three routes as defined in the [Protect API Endpoints](/quickstart/backend/java-spring-security5/interactive/#configure-the-resource-server) section. For the simplicity of this sample we will allow all origins through `@CrossOrigin` annotation. Real applications should configure `CORS` for their use case.

Reviewer 2: ```suggestion
Create a new class named `APIController` to handle requests to the endpoints. The `APIController` has three routes as defined in the [Protect API Endpoints](/quickstart/backend/java-spring-security5/interactive/#configure-the-resource-server) section. For this example, allow all origins through `@CrossOrigin` annotation. Real applications should configure `CORS` for their use case.
```

## Run the application {{{ data-action=code data-code="APIController.java" }}}

To build and run the sample project, execute the `bootRun` Gradle task.

Linux or macOS:

```bash
./gradlew bootRun
```

Windows:

```bash
gradlew.bat bootRun
```

If you are configuring your own application using Maven and the [Spring Boot Maven Plugin](https://docs.spring.io/spring-boot/docs/current/reference/html/build-tool-plugins-maven-plugin.html), you can execute the `spring-boot:run` goal.

Linux or macOS:

```bash
mvn spring-boot:run
```

Windows:

```bash
mvn.cmd spring-boot:run
```

::::checkpoint

:::checkpoint-default

> The sample application will be available at `http://localhost:3010/`. Read about how to test and use your API in the [Using Your API](/quickstart/backend/java-spring-security5/02-using) article.

Reviewer 2: ```suggestion
Navigate to the sample application at `http://localhost:3010/`. Read about how to test and use your API in the [Using Your API](/quickstart/backend/java-spring-security5/02-using) article.

Reviewer: I dont think we should reference the original quickstart here @poovamraj , both the original (02-using), and the suggestion "03-troubleshooting".

Reviewer 2: Updated

:::

:::checkpoint-failure

> Sorry about that. Here's a couple things to double check:

Reviewer 2: ```suggestion
If your application did not launch successfully:
```

* Use the [Troubleshooting](/quickstart/backend/java-spring-security5/03-troubleshooting) section to check your configuration.

Still having issues? Check out our [documentation](https://auth0.com/docs) or visit our [community page](https://community.auth0.com) to get more help.

> :::

Reviewer: ```suggestion
:::
::::
```

## Create the API controller {{{ data-action=code data-code="APIController.java" }}}

Create a new class named `APIController` to handle requests to the endpoints. The `APIController` has three routes as defined in the [Protect API Endpoints](/quickstart/backend/java-spring-security5/interactive/#configure-the-resource-server) section. For this example, allow all origins through `@CrossOrigin` annotation. Real applications should configure `CORS` for their use case.

## Run the application {{{ data-action=code data-code="APIController.java" }}}

To build and run the sample project, execute the `bootRun` Gradle task.

Linux or macOS:

```bash
./gradlew bootRun
```

Windows:

```bash
gradlew.bat bootRun
```

If you are configuring your own application using Maven and the [Spring Boot Maven Plugin](https://docs.spring.io/spring-boot/docs/current/reference/html/build-tool-plugins-maven-plugin.html), you can execute the `spring-boot:run` goal.

Linux or macOS:

```bash
mvn spring-boot:run
```

Windows:

```bash
mvn.cmd spring-boot:run
```

::::checkpoint

:::checkpoint-default

> The sample application will be available at `http://localhost:3010/`. Read about how to test and use your API in the [Using Your API](/quickstart/backend/java-spring-security5/02-using) article.

Reviewer: @poovamraj Should we still refer to the old quickstart here (`/02-using`)?

Author: @frederikprijck you suggest migrating them? We have sub articles
- https://auth0.com/docs/quickstart/backend/java-spring-security5/02-using
- https://auth0.com/docs/quickstart/backend/java-spring-security5/03-troubleshooting

And I don't think we can make it fit in our current design
